### PR TITLE
Make functions attribute in protocols private

### PR DESCRIPTION
**Pull request by eksperimental**

### CODE REVIEW

File: lib/elixir/lib/protocol.ex [hunk 1]
```diff
@@ -275,7 +275,7 @@ defmodule Protocol do
       name = unquote(name)
       arity = unquote(arity)
 
-      @functions [{name, arity} | @functions]
+      @_functions [{name, arity} | @_functions]
```
Comment (josevalim) [inline lib/elixir/lib/protocol.ex:278]:
```suggestion
      @__functions__ [{name, arity} | @__functions__]
```

File: lib/elixir/lib/protocol.ex [hunk 2]
```diff
@@ -706,7 +706,7 @@ defmodule Protocol do
         @compile :debug_info
 
         # Set up a clear slate to store defined functions
-        @functions []
+        @_functions []
```
Comment (josevalim) [inline lib/elixir/lib/protocol.ex:709]:
```suggestion
        @__functions__ []
```

File: lib/elixir/lib/protocol.ex [hunk 3]
```diff
@@ -753,7 +753,7 @@ defmodule Protocol do
     # Callbacks
     callback_metas = callback_metas(env.module, :callback)
     callbacks = :maps.keys(callback_metas)
-    functions = Module.get_attribute(env.module, :functions)
+    functions = Module.get_attribute(env.module, :_functions)
```
Comment (josevalim) [inline lib/elixir/lib/protocol.ex:756]:
```suggestion
    functions = Module.get_attribute(env.module, :__functions__)
```

File: lib/elixir/lib/protocol.ex [hunk 4]
```diff
@@ -883,11 +883,11 @@ defmodule Protocol do
 
       @doc false
       @spec __protocol__(:module) :: __MODULE__
-      @spec __protocol__(:functions) :: unquote(Protocol.__functions_spec__(@functions))
+      @spec __protocol__(:functions) :: unquote(Protocol.__functions_spec__(@_functions))
```
Comment (josevalim) [inline lib/elixir/lib/protocol.ex:886]:
```suggestion
      @spec __protocol__(:functions) :: unquote(Protocol.__functions_spec__(@__functions__))
```

File: lib/elixir/lib/protocol.ex [hunk 5]
```diff
@@ -883,11 +883,11 @@ defmodule Protocol do
...
       @spec __protocol__(:consolidated?) :: boolean
       @spec __protocol__(:impls) :: :not_consolidated | {:consolidated, [module]}
       Kernel.def(__protocol__(:module), do: __MODULE__)
-      Kernel.def(__protocol__(:functions), do: unquote(:lists.sort(@functions)))
+      Kernel.def(__protocol__(:functions), do: unquote(:lists.sort(@_functions)))
```
Comment (josevalim) [inline lib/elixir/lib/protocol.ex:890]:
```suggestion
      Kernel.def(__protocol__(:functions), do: unquote(:lists.sort(@__functions__)))
```

**Comment by josevalim** [comment]
Thanks! I decided to go with `@__functions__` after all because we already use `@__impl__` in this module. :)

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Comment by eksperimental** [comment]
> Thanks! I decided to go with `@__functions__` after all because we already use `@__impl__` in this module. :)

No worries

**Pull request merged**